R-package/src/cmake/modules/FindLibR.cmake

```diff
@@ -0,0 +1,199 @@
+#
+# Defines the following:
+#  LIBR_FOUND
+#  LIBR_HOME
+#  LIBR_EXECUTABLE
+#  LIBR_INCLUDE_DIRS
+#  LIBR_LIB_DIR
+#  LIBR_CORE_LIBRARY
+# and a CMake function to create R.lib for MSVC
+
+if(NOT R_ARCH)
+  if("${CMAKE_SIZEOF_VOID_P}" STREQUAL "4")
+    set(R_ARCH "i386")
+  else()
+    set(R_ARCH "x64")
+  endif()
+endif()
+
+if(NOT ("${R_ARCH}" STREQUAL "x64"))
+  message(FATAL_ERROR "LightGBM's R package currently only supports 64-bit operating systems")
+endif()
+
+# Creates R.lib and R.def in the build directory for linking with MSVC
+function(create_rlib_for_msvc)
+
+  message("Creating R.lib and R.def")
+
+  # various checks and warnings
+  if(NOT WIN32 OR NOT MSVC)
+    message(FATAL_ERROR "create_rlib_for_msvc() can only be used with MSVC")
+  endif()
+
+  if(NOT EXISTS "${LIBR_LIB_DIR}")
+    message(FATAL_ERROR "LIBR_LIB_DIR, '${LIBR_LIB_DIR}', not found")
+  endif()
+
+  find_program(GENDEF_EXE gendef)
+  find_program(DLLTOOL_EXE dlltool)
+
+  if(NOT GENDEF_EXE OR NOT DLLTOOL_EXE)
+    message(FATAL_ERROR "Either gendef.exe or dlltool.exe not found!\
+      \nDo you have Rtools installed with its MinGW's bin/ in PATH?")
+  endif()
+
+  # extract symbols from R.dll into R.def and R.lib import library
+  execute_process(COMMAND ${GENDEF_EXE}
+    "-" "${LIBR_LIB_DIR}/R.dll"
+    OUTPUT_FILE "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+  )
+  execute_process(COMMAND ${DLLTOOL_EXE}
+    "--input-def" "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+    "--output-lib" "${CMAKE_CURRENT_BINARY_DIR}/R.lib"
+  )
+endfunction(create_rlib_for_msvc)
+
+# R version information is used to search for R's libraries in
+# the registry on Windows. Since this code is orchestrated by
+# an R script (src/install.libs.R), that script uses R's built-ins to
+# find the version of R and pass it through as a CMake variable
+if(CMAKE_R_VERSION)
+  message("R version passed into FindLibR.cmake: ${CMAKE_R_VERSION}")
+else()
+  message(FATAL_ERROR "Expected CMAKE_R_VERSION to be passed in but none was provided. Check src/install.libs.R")
+endif()
+
+# Find R executable
+if(APPLE)
+
+  find_library(LIBR_LIBRARIES R)
+
+  if(LIBR_LIBRARIES MATCHES ".*\\.framework")
+    set(LIBR_HOME "${LIBR_LIBRARIES}/Resources")
+    set(LIBR_EXECUTABLE "${LIBR_HOME}/R")
+  else()
+    get_filename_component(_LIBR_LIBRARIES "${LIBR_LIBRARIES}" REALPATH)
+    get_filename_component(_LIBR_LIBRARIES_DIR "${_LIBR_LIBRARIES}" DIRECTORY)
+    set(LIBR_EXECUTABLE "${_LIBR_LIBRARIES_DIR}/../bin/R")
+  endif()
+
+# Unix
+elseif(UNIX)
+
+  # attempt to find R executable
+  if(NOT LIBR_EXECUTABLE)
+
+    # CRAN may run RD CMD CHECK instead of R CMD CHECK,
+    # which can lead to this infamous error:
+    # 'R' should not be used without a path -- see par. 1.6 of the manual
+    find_program(
+      LIBR_EXECUTABLE
+      NO_DEFAULT_PATH
+      HINTS "${CMAKE_CURRENT_BINARY_DIR}" "/usr/bin" "/usr/lib/" "/usr/local/bin/"
+      NAMES R R.exe
+    )
```

Comment:
Hmm, I remember I was typing this comment but it was not included in ,the review... OK, retyping it again.

I think we can do it before any actions for finding R with the help of  `LIBR_HOME` or `LIBR_LIBRARIES`. Also, I don't think that `NAME` `R.exe` will be useful for Unix 😃

Reply:
😬 ha R.exe, bad copy-pasting

Can you clarify what you mean by "before any actions for finding R with the help of LIBR_HOME or LIBR_LIBRARIES"? Do you mean this should be run before `if(APPLE)` (so outside of all the OS-specific checks)?

Reply:
> Do you mean this should be run before if(APPLE) (so outside of all the OS-specific checks)?

Yeah, exactly! I think as a first thing we can simply try to find R via `find_command()` for all OSes before any OS-specific hints. Like
```
if(NOT LIBR_EXECUTABLE)
  find_program( 
        LIBR_EXECUTABLE
        NAMES R R.exe
      )
endif()

if(NOT LIBR_EXECUTABLE)
  if(APPLE)
    ...
endif()
```

Reply:
@StrikerRUS so the reason I _don't_ have that currently is related to  https://github.com/microsoft/LightGBM/pull/2901#discussion_r391733668

When I just had

```
if(NOT LIBR_EXECUTABLE)
  find_program( 
        LIBR_EXECUTABLE
        NAMES R R.exe
      )
endif()
```

During `R  CMD check` it was finding the weird not-exactly-R R executable that is used by CRAN during checks, which results in the error

> 'R' should not be used without a path -- see par. 1.6 of the manual

I didn't see any option [in find_program()](https://cmake.org/cmake/help/v3.12/command/find_program.html) to exclude certain paths (like those that contain `.Rcheck`) so this thing with hints was my workaround until we can come up with something better. 

I guess I could probably do this and then add code that says "if you found the .Rcheck/ one, unset `LIBR_EXECUTABLE` and continue", but I don't see a clean way to say "use  `find_program()` and ignore the R executable that is used by `R CMD check`".

Reply:
Can `NO_PACKAGE_ROOT_PATH` help?

Also, I think that it can be done in the following way. Or am I missing something?..
```
if(NOT LIBR_EXECUTABLE)
  find_program( 
        LIBR_EXECUTABLE
        NAMES R R.exe
      )
  # workaround for R CMD check
  if(LIBR_EXECUTABLE MATCHES ".*lightgbm\\.Rcheck.*")
    unset(LIBR_EXECUTABLE)
  endif()
endif()

if(NOT LIBR_EXECUTABLE)
  if(APPLE)
    ...
endif()
```
https://cmake.org/cmake/help/v3.0/command/unset.html

Reply:
Yep! this is exactly what I meant (I think I was not clear) by:

> I guess I could probably do this and then add code that says "if you found the .Rcheck/ one, unset LIBR_EXECUTABLE and continue", but I don't see a clean way to say "use find_program() and ignore the R executable that is used by R CMD check".

So basically we won't get "the second thing `find_program()` would have found" if the first one is the .Rcheck one. We'll have to fail back to the OS-specific expectations about directory structure. Honestly that's not too bad. I think I'll do that.

as for `NO_PACKAGE_ROOT`...I don't think it will matter for us since we won't have specified `R_ROOT`.

Reply:
wait now I think I understand `NO_PACKAGE_ROOT`!  Ok let me try...

Reply:
Ok I ended up not needing  `NO_PACKAGE_ROOT`. @StrikerRUS whenever you have a chance (thanks by the way for the thorough review!) take a look at my attempt here:

https://github.com/microsoft/LightGBM/pull/2901/commits/5b60e8e7ced86e68cab9b374431ebe846c19d48d

Moving that block up, I had to account for one more special case besides R  CMD CHECK. On Mac, the built-in R GUI is bundled as [R.app](https://cran.r-project.org/bin/macosx/RMacOSX-FAQ.html#R_002eapp) and the executable is (regrettably), also called R. Something like `/Applications/R.app/Contents/MacOS/R`. So that needs to be removed too.

I don't _think_ we'll have a similar issue on Windows ([where you have `Rterm.exe` and `Rgui.exe`](https://cran.r-project.org/bin/windows/base/rw-FAQ.html#How-do-I-run-it_003f)).

Reply:
Thank you for the investigation!

> Ok I ended up not needing `NO_PACKAGE_ROOT`

It's cool! Because `NO_PACKAGE_ROOT_PATH` was introduced only in CMake 3.12.

Reply:
oh nice, ok! I also decided to add https://github.com/microsoft/LightGBM/pull/2901/commits/c9dd50d8bbb4d26b672ccdde7f57b5ea9afa8466, which ensures that all the uses of  `LIBR_EXECUTABLE` ignore environment variables and any `.Rprofile` files.

Just proposed the same to `xgboost`: https://github.com/dmlc/xgboost/pull/5438

Reply:
I think you can contribute to RStudio as well where XGBoost borrowed a lot from.

https://github.com/rstudio/rstudio/blob/master/cmake/modules/FindLibR.cmake

https://github.com/dmlc/xgboost/blob/1de36cdf1e9ff1c93ec648de91ecce288d5430a2/cmake/modules/FindLibR.cmake#L2